cmd/hashers/argon2/calibrate.go

```diff
@@ -0,0 +1,260 @@
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %s of memory\n", currentDuration, config.getMemFormat())
+				}
+
+				if currentDuration < desiredDuration {
+					break
+				}
+				if config.c.Memory <= toKB(adjustMemory) {
+					// adjusting the memory would now result in <= 0B
+					adjustMemory = adjustMemory >> 1
+				}
+				// adjust config
+				config.c.Memory -= toKB(adjustMemory)
+			}
+			if verbose {
+				_, _ = resultColor.Fprintf(cmd.ErrOrStderr(), "Settled on %s of memory.\n", config.getMemFormat())
+				fmt.Fprintf(cmd.ErrOrStderr(), "Increasing iterations to get over %s:\n", desiredDuration)
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %d iterations\n", currentDuration, config.c.Iterations)
+				}
+
+				if currentDuration > desiredDuration {
+					config.c.Iterations -= 1
+					break
+				}
+				// adjust config
+				config.c.Iterations += 1
+			}
+
+			if verbose {
+				fmt.Fprintf(cmd.ErrOrStderr(), "Decreasing iterations to get under %s:\n", desiredDuration)
+			}
+			for {
+				currentDuration, err = probe(cmd, hasher, runs, verbose)
+				if err != nil {
+					return err
+				}
+				if verbose {
+					fmt.Fprintf(cmd.ErrOrStderr(), "  took %s with %d iterations\n", currentDuration, config.c.Iterations)
+				}
+
+				// break also when iterations is 1; this catches the case where 1 was only slightly under the desired time and took longer a bit longer on another run
+				if currentDuration < desiredDuration || config.c.Iterations == 1 {
+					break
+				}
+				// adjust config
+				config.c.Iterations -= 1
+			}
+			if verbose {
+				_, _ = resultColor.Fprintf(cmd.ErrOrStderr(), "Settled on %d iterations.\n\n", config.c.Iterations)
+			}
+
+			e := json.NewEncoder(cmd.OutOrStdout())
+			e.SetIndent("", "  ")
+			return e.Encode(config.c)
+		},
+	}
+
+	registerArgon2Flags(cmd.Flags())
+	cmd.Flags().BoolP(FlagVerbose, "v", false, "verbose output")
+	cmd.Flags().IntP(FlagRuns, "r", 2, "runs per probe, median of all runs is taken as the result")
+
+	return cmd
+}
+
+func toKB(b datasize.ByteSize) uint32 {
+	return uint32(b / datasize.KB)
+}
+
+func probe(cmd *cobra.Command, hasher hash.Hasher, runs int, verbose bool) (time.Duration, error) {
+	start := time.Now()
+
+	var mid time.Time
+	for i := 0; i < runs; i++ {
+		mid = time.Now()
+		_, err := hasher.Generate([]byte("password"))
+		if err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Could not generate a hash: %s\n", err)
+			return 0, cmdx.FailSilently(cmd)
+		}
+		if verbose {
+			fmt.Fprintf(cmd.OutOrStdout(), "    took %s in try %d\n", time.Since(mid), i)
+		}
+	}
+
+	return time.Duration(int64(time.Since(start)) / int64(runs)), nil
+}
+
+func registerArgon2Flags(flags *pflag.FlagSet) {
+	flags.StringP(FlagStartMemory, "m", "4GB", "amount of memory to start probing at")
```

Comment:
Would it make sense to limit this to the maximum RAM available or 4GB if more than 4GB are available?

Reply:
I could not find any way to figure out how much RAM is available.... except for using cgo and doing linux syscalls
which obviously only works on linux

Reply:
Ahh I thought https://golang.org/pkg/runtime/#MemStats could solve this but it looks like it only returns info about the running Go process.

Guess this doesn't work then!